### ISSUE TITLE
PromiseProxy deprecation can trigger infinite recursion in ember-concurrency

### ISSUE THREAD

**Issue by runspired** [emberjs/data#8312]
Related to #7997 

Reported by several in slack.

Easy work around is to wrap or extract the `await save()` or `await findRecord()` calls such that they aren't the promise being seen by ember-concurrency.

**Comment by pernambucano** [comment]
I was receiving another error "Cannot read property 'bind' of undefined" while trying to save inside a task on ember 4.4, and this workaround also work for it. Thanks!

**Comment by runspired** [comment]
Note: I'd be happy to help fix this but I don't have an app that reproduces this to debug the root cause. If someone is able to pair ping me in discord.

**Comment by cddso** [comment]
I made a very simple app that showcases the infinite recursion problem. There's hardly any setup.

https://github.com/cddso/ember-data-concurrency-app

The latest commit (https://github.com/cddso/ember-data-concurrency-app/commit/c81ef5d1b28bc79e3f4a1f39100140e637d44ec4) is the entire setup.

**Comment by cah-brian-gantzler** [comment]
There was an issue similar to this with ember-resources when a store.query was the first yield. Following this issue https://github.com/NullVoxPopuli/ember-resources/issues/340 I added an additional yield and everything was working fine until ember data 4.8.   Thought you might want to read up on that as it may be related and help you find the cause

**Comment by esbanarango** [comment]
Facing this same issue when upgrading from `ember-data@4.7.3` to `ember-data@4.9.1`:

This will throw an infinite recursion error:
<img width="1547" alt="SignalUi" src="https://user-images.githubusercontent.com/1156865/210459295-51070d7b-1146-43b9-80f0-e45fe5fff0d4.png">


#### Example to reproduce

```ts
  // ERROR
  loadDataTask = task({ restartable: true }, async () => {
    this.data = await this.store.query('my-model')
  });
```

**No error**
```ts
  // NO error
  loadDataTask = task({ restartable: true }, async () => {
    await this.store
      .query('my-model')
      .then((data) => {
        this.data = data;
      });
  });
```

**Comment by cah-brian-gantzler** [comment]
Checking if any progress has been made on this. Stuck on 4.7, starting to get a few version behind

**Comment by runspired** [comment]
@cah-brian-gantzler we're stuck too :) so I *will* get around to it very soon <3

**Comment by cah-brian-gantzler** [comment]
Thats great. Thanks. 

Dont understand the fix, but seems it was pretty simple.

Since you back ported it to 4.8 (because it was an LTS) doesnt this mean that 4.9 and 4.10 dont have it? That could be confusing to someone who updated a version at a time.

**Comment by bardzusny** [comment]
We've stumbled upon this issue when doing a major 3->4 upgrade in a large app. Upgrading directly to 4.8 seemed like a lost effort because of multiple other changes - but here's a sensible workaround that let us avoid this bug while targeting 4.4:

First override the original DS Model class somewhere:
```ts
// app/models/base.ts

import Model from '@ember-data/model';
import RSVP from 'rsvp';

// Wrap async model methods in RSVP.resolve(...) to workaround a bug in Ember Data 4.4 causing errors when
// they're called in the context of ember-concurrency tasks: https://github.com/emberjs/data/issues/8312
//
// (It's fixed in Ember Data  4.8, so it can be removed once we upgrade to it.)
class FixedModel extends Model implements Model {
  save(_opts: { adapterOptions?: object | undefined }) {
    return RSVP.resolve(super.save(...arguments));
  }

  reload(_opts: { adapterOptions?: object | undefined }) {
    return RSVP.resolve(super.reload(...args));
  }

  destroyRecord(_opts: { adapterOptions?: object | undefined }) {
    return RSVP.resolve(super.destroyRecord(...args));
  }

}

export default FixedModel;

```

Then extend your actual models from it:
```ts
// app/models/foo.ts

import Model from './base';

export default FooModel extends Model {
  ...
}
